[ActionSheet] Add ability to add a divider between header and table

codeman7 opened the pull request:
This PR adds an API 

*new property* `*headerDividerColor` on _MDCActionSheetController_
*new property* `*showHeaderDivider` on _MDCActionSheetController_

that will allow clients to add a divider to the action sheet between the header and the table. If a client wants to show a header divider they would need to set `showHeaderDivider` and then set an appropriate color for their use case, by default it is `clearColor`. By default the property `showHeaderDivider` view is `NO` so that we do not break clients. This matches a pattern outlined in [`MDCBannerView`](https://github.com/material-components/material-components-ios/blob/ac114b4bda9dbad328ae445c4529a43a94ccd97d/components/Banner/src/MDCBannerView.h#L76-L88).

Closes #7395

codeman7 committed:
initial

codeman7 committed:
fix the :bug:

codeman7 committed:
fix :camera:

material-automation commented on components/ActionSheet/src/MDCActionSheetController.h:
```diff
@@ -203,6 +203,15 @@ __attribute__((objc_subclassing_restricted)) @interface MDCActionSheetController
 
+/**
+ The color of the divider between the header and actions.
+ 
```
clang-format suggested the following change:


```
- 
+
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/src/MDCActionSheetController.h -->

material-automation commented on components/ActionSheet/src/MDCActionSheetController.h:
```diff
@@ -203,6 +203,15 @@ __attribute__((objc_subclassing_restricted)) @interface MDCActionSheetController
+ The color of the divider between the header and actions.
+ 
+ @note Setting this to any value other than @c nil will result in the table being shifted down 1pt.
+ 
```
clang-format suggested the following change:


```
- 
+
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/src/MDCActionSheetController.h -->

material-automation commented on components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m:
```diff
@@ -818,4 +818,76 @@ - (void)testActionSheetWhenEveryActionHasCustomTintColor {
+  MDCActionSheetAction *action1 =
+  [MDCActionSheetAction actionWithTitle:kShortTitle1Latin
+                                  image:[UIImage mdc_testImageOfSize:CGSizeMake(24, 24)]
+                                handler:nil];
```
clang-format suggested the following change:


```
-  [MDCActionSheetAction actionWithTitle:kShortTitle1Latin
-                                  image:[UIImage mdc_testImageOfSize:CGSizeMake(24, 24)]
-                                handler:nil];
+      [MDCActionSheetAction actionWithTitle:kShortTitle1Latin
+                                      image:[UIImage mdc_testImageOfSize:CGSizeMake(24, 24)]
+                                    handler:nil];
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m -->

material-automation commented on components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m:
```diff
@@ -818,4 +818,76 @@ - (void)testActionSheetWhenEveryActionHasCustomTintColor {
+  MDCActionSheetAction *action2 =
+  [MDCActionSheetAction actionWithTitle:kShortTitle2Latin
+                                  image:[UIImage mdc_testImageOfSize:CGSizeMake(24, 24)]
+                                handler:nil];
```
clang-format suggested the following change:


```
-  [MDCActionSheetAction actionWithTitle:kShortTitle2Latin
-                                  image:[UIImage mdc_testImageOfSize:CGSizeMake(24, 24)]
-                                handler:nil];
+      [MDCActionSheetAction actionWithTitle:kShortTitle2Latin
+                                      image:[UIImage mdc_testImageOfSize:CGSizeMake(24, 24)]
+                                    handler:nil];
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m -->

material-automation commented on components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m:
```diff
@@ -818,4 +818,76 @@ - (void)testActionSheetWhenEveryActionHasCustomTintColor {
+                                  image:[UIImage mdc_testImageOfSize:CGSizeMake(24, 24)]
+                                handler:nil];
+  MDCActionSheetController *controller =
+  [MDCActionSheetController actionSheetControllerWithTitle:@"Foo"];
```
clang-format suggested the following change:


```
-  [MDCActionSheetController actionSheetControllerWithTitle:@"Foo"];
+      [MDCActionSheetController actionSheetControllerWithTitle:@"Foo"];
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m -->

material-automation commented on components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m:
```diff
@@ -818,4 +818,76 @@ - (void)testActionSheetWhenEveryActionHasCustomTintColor {
+  [MDCActionSheetController actionSheetControllerWithTitle:@"Foo"];
+  [controller addAction:action1];
+  [controller addAction:action2];
+  
```
clang-format suggested the following change:


```
-  
+
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m -->

material-automation commented on components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m:
```diff
@@ -818,4 +818,76 @@ - (void)testActionSheetWhenEveryActionHasCustomTintColor {
+  // When
+  controller.headerDividerColor = UIColor.blackColor;
+  controller.view.bounds = CGRectMake(0, 0, 320, 200);
+  
```
clang-format suggested the following change:


```
-  
+
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m -->

material-automation commented on components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m:
```diff
@@ -818,4 +818,76 @@ - (void)testActionSheetWhenEveryActionHasCustomTintColor {
+  MDCActionSheetAction *action1 =
+  [MDCActionSheetAction actionWithTitle:kShortTitle1Latin
+                                  image:[UIImage mdc_testImageOfSize:CGSizeMake(24, 24)]
+                                handler:nil];
```
clang-format suggested the following change:


```
-  [MDCActionSheetAction actionWithTitle:kShortTitle1Latin
-                                  image:[UIImage mdc_testImageOfSize:CGSizeMake(24, 24)]
-                                handler:nil];
+      [MDCActionSheetAction actionWithTitle:kShortTitle1Latin
+                                      image:[UIImage mdc_testImageOfSize:CGSizeMake(24, 24)]
+                                    handler:nil];
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m -->

material-automation commented on components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m:
```diff
@@ -818,4 +818,76 @@ - (void)testActionSheetWhenEveryActionHasCustomTintColor {
+  MDCActionSheetAction *action2 =
+  [MDCActionSheetAction actionWithTitle:kShortTitle2Latin
+                                  image:[UIImage mdc_testImageOfSize:CGSizeMake(24, 24)]
+                                handler:nil];
```
clang-format suggested the following change:


```
-  [MDCActionSheetAction actionWithTitle:kShortTitle2Latin
-                                  image:[UIImage mdc_testImageOfSize:CGSizeMake(24, 24)]
-                                handler:nil];
+      [MDCActionSheetAction actionWithTitle:kShortTitle2Latin
+                                      image:[UIImage mdc_testImageOfSize:CGSizeMake(24, 24)]
+                                    handler:nil];
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m -->

material-automation commented:
bazel detected changes to the following targets:

```
//components/ActionSheet:snapshot_tests
//components/ActionSheet:snapshot_tests_test_bundle
//components/ActionSheet:unit_tests
//components/ActionSheet:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/ActionSheet:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/ActionSheet:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/ActionSheet:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/ActionSheet:unit_tests_IPHONE_X_IN_11_0
//components/ActionSheet:unit_tests_IPHONE_X_IN_11_0_test_bundle
```

<!-- identifier: affected-targets -->

material-automation commented on components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m:
```diff
@@ -818,4 +818,76 @@ - (void)testActionSheetWhenEveryActionHasCustomTintColor {
+                                  image:[UIImage mdc_testImageOfSize:CGSizeMake(24, 24)]
+                                handler:nil];
+  MDCActionSheetController *controller =
+  [MDCActionSheetController actionSheetControllerWithTitle:@"Foo" message:@"Bar"];
```
clang-format suggested the following change:


```
-  [MDCActionSheetController actionSheetControllerWithTitle:@"Foo" message:@"Bar"];
+      [MDCActionSheetController actionSheetControllerWithTitle:@"Foo" message:@"Bar"];
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m -->

material-automation commented on components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m:
```diff
@@ -818,4 +818,76 @@ - (void)testActionSheetWhenEveryActionHasCustomTintColor {
+  [MDCActionSheetController actionSheetControllerWithTitle:@"Foo" message:@"Bar"];
+  [controller addAction:action1];
+  [controller addAction:action2];
+  
```
clang-format suggested the following change:


```
-  
+
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m -->

material-automation commented on components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m:
```diff
@@ -818,4 +818,76 @@ - (void)testActionSheetWhenEveryActionHasCustomTintColor {
+  // When
+  controller.headerDividerColor = UIColor.blackColor;
+  controller.view.bounds = CGRectMake(0, 0, 320, 200);
+  
```
clang-format suggested the following change:


```
-  
+
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m -->

material-automation commented on components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m:
```diff
@@ -818,4 +818,76 @@ - (void)testActionSheetWhenEveryActionHasCustomTintColor {
+  MDCActionSheetAction *action1 =
+  [MDCActionSheetAction actionWithTitle:kShortTitle1Latin
+                                  image:[UIImage mdc_testImageOfSize:CGSizeMake(24, 24)]
+                                handler:nil];
```
clang-format suggested the following change:


```
-  [MDCActionSheetAction actionWithTitle:kShortTitle1Latin
-                                  image:[UIImage mdc_testImageOfSize:CGSizeMake(24, 24)]
-                                handler:nil];
+      [MDCActionSheetAction actionWithTitle:kShortTitle1Latin
+                                      image:[UIImage mdc_testImageOfSize:CGSizeMake(24, 24)]
+                                    handler:nil];
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m -->

material-automation commented on components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m:
```diff
@@ -818,4 +818,76 @@ - (void)testActionSheetWhenEveryActionHasCustomTintColor {
+  MDCActionSheetAction *action2 =
+  [MDCActionSheetAction actionWithTitle:kShortTitle2Latin
+                                  image:[UIImage mdc_testImageOfSize:CGSizeMake(24, 24)]
+                                handler:nil];
```
clang-format suggested the following change:


```
-  [MDCActionSheetAction actionWithTitle:kShortTitle2Latin
-                                  image:[UIImage mdc_testImageOfSize:CGSizeMake(24, 24)]
-                                handler:nil];
+      [MDCActionSheetAction actionWithTitle:kShortTitle2Latin
+                                      image:[UIImage mdc_testImageOfSize:CGSizeMake(24, 24)]
+                                    handler:nil];
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m -->

material-automation commented on components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m:
```diff
@@ -818,4 +818,76 @@ - (void)testActionSheetWhenEveryActionHasCustomTintColor {
+                                  image:[UIImage mdc_testImageOfSize:CGSizeMake(24, 24)]
+                                handler:nil];
+  MDCActionSheetController *controller =
+  [MDCActionSheetController actionSheetControllerWithTitle:nil];
```
clang-format suggested the following change:


```
-  [MDCActionSheetController actionSheetControllerWithTitle:nil];
+      [MDCActionSheetController actionSheetControllerWithTitle:nil];
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m -->

material-automation commented on components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m:
```diff
@@ -818,4 +818,76 @@ - (void)testActionSheetWhenEveryActionHasCustomTintColor {
+  [MDCActionSheetController actionSheetControllerWithTitle:nil];
+  [controller addAction:action1];
+  [controller addAction:action2];
+  
```
clang-format suggested the following change:


```
-  
+
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m -->

material-automation commented on components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m:
```diff
@@ -818,4 +818,76 @@ - (void)testActionSheetWhenEveryActionHasCustomTintColor {
+  // When
+  controller.headerDividerColor = UIColor.blackColor;
+  controller.view.bounds = CGRectMake(0, 0, 320, 200);
+  
```
clang-format suggested the following change:


```
-  
+
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m -->

material-automation commented on components/ActionSheet/tests/unit/MDCActionSheetTest.m:
```diff
@@ -457,4 +463,28 @@ - (void)testSettingBaseOverrideBaseElevationReturnsSetValue {
+- (void)testSetHeaderDividerColor {
+  // Given
+  UIColor *fakeColor = UIColor.orangeColor;
+  
```
clang-format suggested the following change:


```
-  
+
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/tests/unit/MDCActionSheetTest.m -->

material-automation commented on components/ActionSheet/tests/unit/MDCActionSheetTest.m:
```diff
@@ -457,4 +463,28 @@ - (void)testSettingBaseOverrideBaseElevationReturnsSetValue {
+  
+  // When
+  self.actionSheet.headerDividerColor = fakeColor;
+  
```
clang-format suggested the following change:


```
-  
+
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/tests/unit/MDCActionSheetTest.m -->

material-automation commented on components/ActionSheet/tests/unit/MDCActionSheetTest.m:
```diff
@@ -457,4 +463,28 @@ - (void)testSettingBaseOverrideBaseElevationReturnsSetValue {
+- (void)testSetHeaderDividerColorThenResetToNil {
+  // Given
+  self.actionSheet.headerDividerColor = UIColor.orangeColor;
+  
```
clang-format suggested the following change:


```
-  
+
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/tests/unit/MDCActionSheetTest.m -->

material-automation commented on components/ActionSheet/tests/unit/MDCActionSheetTest.m:
```diff
@@ -457,4 +463,28 @@ - (void)testSettingBaseOverrideBaseElevationReturnsSetValue {
+  
+  // When
+  self.actionSheet.headerDividerColor = nil;
+  
```
clang-format suggested the following change:


```
-  
+
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/tests/unit/MDCActionSheetTest.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

codeman7 committed:
I :heart: clang

material-automation commented:
## API diff detected the following changes

### ActionSheet

#### MDCActionSheetController

*new* property: `showsHeaderDivider` in `MDCActionSheetController`

*new* property: `headerDividerColor` in `MDCActionSheetController`

<!-- identifier: apidiff -->

fluxminds commented:
Two properties (`showDivider`, `dividerColor`) were used in Banner for clarity. Some context on the discussion can be found in https://github.com/material-components/material-components-ios/pull/7567. Maybe it would be helpful to keep the consistence for users?

romoore commented on components/ActionSheet/src/MDCActionSheetController.m:
```diff
@@ -175,7 +179,14 @@ - (void)viewWillLayoutSubviews {
   CGSize size = [self.header sizeThatFits:CGRectStandardize(self.view.bounds).size];
   self.header.frame = CGRectMake(0, 0, self.view.bounds.size.width, size.height);
-  UIEdgeInsets insets = UIEdgeInsetsMake(self.header.frame.size.height, 0, 0, 0);
+  if (self.title != nil || self.message != nil) {
```
Shouldn't this be checking the `headerDividerColor` to see if it's `nil`? As I understand the header documentation, it's the color that determines if the view is added.

codeman7 replied:
This has been updated the headerDividerView is never nil.

romoore commented on components/ActionSheet/src/MDCActionSheetController.m:
```diff
@@ -175,7 +179,14 @@ - (void)viewWillLayoutSubviews {
   self.header.frame = CGRectMake(0, 0, self.view.bounds.size.width, size.height);
-  UIEdgeInsets insets = UIEdgeInsetsMake(self.header.frame.size.height, 0, 0, 0);
+  if (self.title != nil || self.message != nil) {
+    [self.view addSubview:self.headerDividerView];
```
Won't this re-add the divider view on every layout pass? Can you check to see if the divider view's `superview` is `self.view`? Or maybe to the `addSubview:` and `removeFromSuperview` in the `setHeaderDividerColor:` method?

codeman7 replied:
This has been removed.

romoore commented on components/ActionSheet/tests/unit/MDCActionSheetTest.m:
```diff
@@ -457,4 +463,28 @@ - (void)testSettingBaseOverrideBaseElevationReturnsSetValue {
 
+- (void)testSetHeaderDividerColor {
+  // Given
+  UIColor *fakeColor = UIColor.orangeColor;
```
Tip: Nit: Why is this a fake color?  It looks like a real color.

codeman7 replied:
It was just to make sure the color we set is the same color in the expectation, I have changed the name to be `expectedColor` to clarify this more.

romoore commented on components/ActionSheet/tests/unit/MDCActionSheetTest.m:
```diff
@@ -457,4 +463,28 @@ - (void)testSettingBaseOverrideBaseElevationReturnsSetValue {
   XCTAssertEqualWithAccuracy(self.actionSheet.mdc_overrideBaseElevation, fakeElevation, 0.001);
 }
 
+- (void)testSetHeaderDividerColor {
```
Please add a test that setting the `headerDividerColor` causes the bounds/frame to increase by exactly 1 point.

codeman7 replied:
It only changes the contentOffset of the tableView and I have added test for that.

romoore commented on components/ActionSheet/src/MDCActionSheetController.m:
```diff
@@ -175,7 +179,14 @@ - (void)viewWillLayoutSubviews {
   self.header.frame = CGRectMake(0, 0, self.view.bounds.size.width, size.height);
-  UIEdgeInsets insets = UIEdgeInsetsMake(self.header.frame.size.height, 0, 0, 0);
+  if (self.title != nil || self.message != nil) {
+    [self.view addSubview:self.headerDividerView];
```
`-[UIView addSubview:]` requires a `nonnull` argument and it looks like `headerDividerView` can be `nil` at this point if the divider color is `nil`.

codeman7 replied:
The headerDividerView is no nonnull.

romoore reviewed (commented):
The documentation and implementation don't seem to match. The docs suggest that the divider view is fully controlled by the color, but the implementation seems to rely on a title or message to exist.  It also appears like the code is potentially passing `nil` to the `nonnull` `addSubview:` API.

romoore reviewed (dismissed):
Oops, meant to Request Changes. Please see my previous review comments.

codeman7 committed:
clean

codeman7 committed:
feedback

codeman7 committed:
cleanup

codeman7 committed:
cleanup v2

codeman7 committed:
fix :bug:

codeman7 committed:
clenaup

codeman7 committed:
cleanup

codeman7 committed:
final cleanup

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

material-automation commented on components/ActionSheet/tests/unit/MDCActionSheetTest.m:
```diff
@@ -457,4 +464,76 @@ - (void)testSettingBaseOverrideBaseElevationReturnsSetValue {
+
+  // Then
+  XCTAssertEqualWithAccuracy(self.actionSheet.tableView.contentOffset.y,
+                             originalTableContentInsets, 0.001);
```
clang-format suggested the following change:


```
-  XCTAssertEqualWithAccuracy(self.actionSheet.tableView.contentOffset.y,
-                             originalTableContentInsets, 0.001);
+  XCTAssertEqualWithAccuracy(self.actionSheet.tableView.contentOffset.y, originalTableContentInsets,
+                             0.001);
```

<!-- identifier: clang-format -->
<!-- File: components/ActionSheet/tests/unit/MDCActionSheetTest.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

codeman7 committed:
clang

codeman7 committed:
fix :bug:

romoore commented on components/ActionSheet/src/MDCActionSheetController.h:
```diff
@@ -203,6 +203,21 @@ __attribute__((objc_subclassing_restricted)) @interface MDCActionSheetController
+
+ Defaults to UIColor.clearColor.
+ */
+@property(nonatomic, copy, nonnull) UIColor *headerDividerColor;
```
I don't really understand why there are two separate properties controlling the visibility for this view. Could you help me understand why the two properties are necessary?

codeman7 replied:
This is following the convention set in #7567, I think this is what @wenyuzhang666 and @jverkoey would prefer.

romoore replied:
While I don't believe the boolean is strictly necessary, in #7567 the default value is visible. It's documented as "light grey". In this case it appears the default color is clear.  In order to use the property on Banner the user only needs to toggle the boolean flag. However here, they need to toggle the boolean *and* set a non-clear color. I think there needs to be a visible default color for the divider so users don't have to take two steps to enable the divider.

I'd also like to note that in [the relevant conversation thread for #7567](https://github.com/material-components/material-components-ios/pull/7567/files#r293004510), having both properties is referred to as an "acceptable path forward". This would seem like there was a compromise made at some point and the final decision was a result of that compromise.  Unfortunately the reasons for the decision aren't entirely clear, except a reference to theming extensions. I would be cautious about using that decision as the only basis for this one.

codeman7 replied:
I've added a default color to match Banner.

romoore commented on components/ActionSheet/src/MDCActionSheetController.h:
```diff
@@ -203,6 +203,21 @@ __attribute__((objc_subclassing_restricted)) @interface MDCActionSheetController
+
+ Defaults to NO.
+ */
+@property(nonatomic, assign) BOOL showHeaderDivider;
```
Based on the code, this only has any effect if there is either a title or message (or both).  Please update the documentation (or implementation) to reflect the correct behavior.

codeman7 replied:
I didn't really like the _magic_ I had added in that situation. I think its better just to have clients have full control over showing the headerDivider or not regardless of the title and message being set. This has been cleaned up.

romoore commented on components/ActionSheet/src/MDCActionSheetController.h:
```diff
@@ -203,6 +203,21 @@ __attribute__((objc_subclassing_restricted)) @interface MDCActionSheetController
+
+ Defaults to NO.
+ */
+@property(nonatomic, assign) BOOL showHeaderDivider;
```
Nit/naming: I believe `showsHeaderDivider` or `headerDividerVisible` would better align with Apple's naming conventions.
*  `setShowsHeaderDivider` (setter) and `showsHeaderDivider` (getter)
*  `setHeaderDividerVisible` (setter) and `isHeaderDividerVisible` (getter)

From [Apple's Naming guidelines](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingMethods.html), "If the property is expressed as a verb, …the verb should be in the simple present tense."

codeman7 replied:
This has been updated.

romoore commented on components/ActionSheet/src/MDCActionSheetController.m:
```diff
@@ -393,6 +408,12 @@ - (UIColor *)messageTextColor {
 }
 
+- (void)setHeaderDividerColor:(UIColor *)headerDividerColor {
+  _headerDividerColor = [headerDividerColor copy];
```
Tip: If the `headerDividerView` is `nonnull`, can this just be a pass-through property?

codeman7 replied:
Whoops, forgot to clean this up after addressing other things, thanks for catching this!

romoore commented on components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m:
```diff
@@ -871,4 +871,55 @@ - (void)testThreeActionsSufficientSizeShortTextLTRWithDefaultPresentationStyleOn
   [self snapshotVerifyViewForIOS13:window];
 }
 
+- (void)testActionSheetWithHeaderShown {
```
Should there be an iOS 13 dynamic color test? Will that be done in a follow-up for the theming extension?

codeman7 replied:
I don't believe so. I think we only test that dynamic colors are supported for `CGColor` instances but this is setting a UIColor on a UIView so I think we are okay.

romoore replied:
Acknowledged.

codeman7 committed:
feedback

codeman7 committed:
cleanup

codeman7 committed:
cleanup

romoore commented on components/ActionSheet/src/MDCActionSheetController.m:
```diff
@@ -393,6 +407,23 @@ - (UIColor *)messageTextColor {
+}
+
+- (void)setShowsHeaderDivider:(BOOL)showsHeaderDivider {
+  if (_showsHeaderDivider == showsHeaderDivider) {
```
Tip: Can this be bound to the [`hidden` property on the view](https://developer.apple.com/documentation/uikit/uiview/1622585-hidden)? I believe that property will take care of any layout/display updates required.

codeman7 replied:
This has been updated.

romoore commented on components/ActionSheet/src/MDCActionSheetController.m:
```diff
@@ -175,7 +181,15 @@ - (void)viewWillLayoutSubviews {
   CGSize size = [self.header sizeThatFits:CGRectStandardize(self.view.bounds).size];
   self.header.frame = CGRectMake(0, 0, self.view.bounds.size.width, size.height);
-  UIEdgeInsets insets = UIEdgeInsetsMake(self.header.frame.size.height, 0, 0, 0);
+  CGFloat dividerHeight = 1;
```
Tip: I'm curious about this block of logic.  Is there a reason the same frame cannot be used in both cases?  If the height is zero, then it seems it will have the same effect (not resizing the view) of setting the frame to `CGRectZero`.

```objc
CGFloat dividerHeight = self.showsHeaderDivider ? 1 : 0;
self.headerDividerView.frame =
        CGRectMake(0, size.height, CGRectGetWidth(self.view.bounds), dividerHeight);
UIEdgeInsets insets = UIEdgeInsetsMake(size.height + dividerHeight, 0, 0, 0);
```

codeman7 replied:
Ah much cleaner, I didn't even consider that, thanks!

romoore commented on components/ActionSheet/src/MDCActionSheetController.m:
```diff
@@ -393,6 +407,23 @@ - (UIColor *)messageTextColor {
 
+- (void)setHeaderDividerColor:(UIColor *)headerDividerColor {
+  self.headerDividerView.backgroundColor = headerDividerColor;
+  [self.view setNeedsDisplay];
```
Tip: Is this call necessary? I'm surprised that setting the `backgroundColor` isn't causing a layout pass. When I attempted to reproduce the need for `setNeedsDisplay`, the subview's updated `backgroundColor` is shown immediately.

At the least, please limit forced display updates to the headerDividerView's frame:

```objc
// Since the frame is a computed property, rely on the divider's bounds.
[self.view setNeedsDisplayInRect:[self.view convertRect:self.headerDividerView.bounds 
                                               fromView:self.headerDividerView]];
```

codeman7 replied:
This is no longer needed since we don't need `setNeedsDisplay`

romoore commented on components/ActionSheet/tests/snapshot/MDCActionSheetControllerSnapshotTests.m:
```diff
@@ -871,4 +871,55 @@ - (void)testThreeActionsSufficientSizeShortTextLTRWithDefaultPresentationStyleOn
+  [self generateSnapshotAndVerifyForView:self.actionSheetController.view];
+}
+
+- (void)testActionSheetWithHeaderShownButNoTitleOrMessage {
```
Looks like the golden image for this test needs updating.

codeman7 replied:
done

codeman7 committed:
feedback

codeman7 committed:
clang

codeman7 committed:
Remove setNeedsDisplay

Still need to update the snapshots and test if setNeedsDisplay is needed. This is a commit right before client teams app built, more of a stash may need to revert

codeman7 committed:
update golden

codeman7 committed:
fix bug

codeman7 committed:
remove record mode

codeman7 committed:
feedback default color

romoore commented on components/ActionSheet/src/MDCActionSheetController.h:
```diff
@@ -203,6 +203,22 @@ __attribute__((objc_subclassing_restricted)) @interface MDCActionSheetController
+@property(nonatomic, assign) BOOL showsHeaderDivider;
+
+/**
+ The color of the divider between the header and actions. This API should be used with
```
Nit/spelling: s/visable/visible/ s/ligh/light/

codeman7 replied:
Updated

romoore commented on components/ActionSheet/src/MDCActionSheetController.h:
```diff
@@ -203,6 +203,22 @@ __attribute__((objc_subclassing_restricted)) @interface MDCActionSheetController
+@property(nonatomic, assign) BOOL showsHeaderDivider;
+
+/**
+ The color of the divider between the header and actions. This API should be used with
```
Tip: You can probably remove the sentence, "This API should be used..." since I believe it's no longer required that updating this property is necessary to make the divider visible.

codeman7 replied:
Removed.

romoore commented on components/ActionSheet/src/MDCActionSheetController.m:
```diff
@@ -23,6 +23,7 @@
 static NSString *const kReuseIdentifier = @"BaseCell";
 static const CGFloat kActionImageAlpha = (CGFloat)0.6;
 static const CGFloat kActionTextAlpha = (CGFloat)0.87;
+static const CGFloat kDividerDefaultOpacity = (CGFloat)0.12;
```
Nit: Please name this `kDividerDefaultAlpha` to match the names of similar constants above.

codeman7 replied:
Updated, good catch!

romoore commented on components/ActionSheet/src/MDCActionSheetController.h:
```diff
@@ -203,6 +203,22 @@ __attribute__((objc_subclassing_restricted)) @interface MDCActionSheetController
+ The color of the divider between the header and actions. This API should be used with
+ @c showsHeaderDivider to make the divider visable.
+
+ Defaults to a ligh grey color.
```
Either this comment or the code below is incorrect. The color is not a light grey color, but a partially-transparent black color.  "Light grey" generally means an opaque white color with a white value greater than 0.5.

Perhaps `[UIColor colorWithWhite:0.88 alpha:1]` or something similar?  As an added benefit, opaque colors often render faster because no blending has to take place.

codeman7 replied:
Updated the docs.

romoore commented on components/ActionSheet/src/MDCActionSheetController.m:
```diff
@@ -131,6 +135,10 @@ - (instancetype)initWithTitle:(NSString *)title message:(NSString *)message {
+    _headerDividerView = [[UIView alloc] init];
+    _headerDividerView.backgroundColor =
+        [UIColor.blackColor colorWithAlphaComponent:kDividerDefaultOpacity];
+    ;
```
Nit: Extra semicolon here.

codeman7 replied:
Done

romoore reviewed (approved):
LGTM, but please correct the default color to be opaque or else update the documentation to reflect the documentation that it's actually a partially-transparent black.

codeman7 committed:
feedback

codeman7 committed:
make it compile
